input: pass modifier state in set_mouse_pos

occivink opened the pull request:
This makes it possible to use CTRL+MOUSE_MOVE and similar bindings

Tested on x11 and wayland. 

For win32 I made the change 'blind', as in I couldn't compile or test it. For cocoa I could not figure how to retrieve the modifiers, so currently they are not passed.

I wasn't sure from reading `client-api-changes.rst` if this change should be mentioned, should it?

Motivation for this change: I have a plugin where the user can draw rectangles on the video with the mouse. I would like to make it similar to other programs (like inkscape), where CTRL can be held to preserve the aspect ratio, or SHIFT to preserve the center (or both)

occivink committed:
input: pass modifier state in set_mouse_pos

This makes it possible to use CTRL+MOUSE_MOVE and similar bindings

Akemi commented:
not sure, but shouldn't this be handled like the calc geometry functions, eg when  function parameter is added we keep the old function that calls the new one with a default value?
https://github.com/mpv-player/mpv/blob/c2dec023b79ebb54160b5a7a1c5f2b609627f167/video/out/win_state.h#L26-L32
https://github.com/mpv-player/mpv/blob/c2dec023b79ebb54160b5a7a1c5f2b609627f167/video/out/win_state.c#L80-L145

Dudemanguy commented:
>I wasn't sure from reading client-api-changes.rst if this change should be mentioned, should it?

This isn't a part of the client api so no need to worry about that. It's technically an interface change though (you can now bind modifiers to mouse movements), so you could mention it there.

>not sure, but shouldn't this be handled like the calc geometry functions, eg when function parameter is added we keep the old function that calls the new one with a default value?

I personally don't think so. In the calc geometry case, there's value in keeping the old functions since not everything needs `vo_calc_window_geometry3`. Honestly, `vo_calc_window_geometry2` probably didn't need to be a separate function anyway and simply adding the DPI as an extra argument and passing 1 if DPI wasn't needed would have been fine but that's an old story. As for this, if the vo is not expected to handle modifiers, just passing 0 is perfectly fine imo since it's only used internally.

Dudemanguy commented on player/command.c:
```diff
@@ -5855,7 +5855,7 @@ static void cmd_mouse(void *p)
     if (pre_key)
         mp_input_put_key_artificial(mpctx->input, pre_key);
-    mp_input_set_mouse_pos_artificial(mpctx->input, x, y);
+    mp_input_set_mouse_pos_artificial(mpctx->input, x, y, 0);
```
Is there any value to allowing this command to pass modifiers? I'm honestly not sure, but I'd thought I'd just mention it.

occivink replied:
I think it would make sense, yes. It occurred to me as well, but then I realized that this command does not allow you to pass modifiers to artificial button clicks either yet, so maybe as a separate enhancement?

Dudemanguy replied:
If the change is just as simple as adding a new argument to this input command, than it should be part of the same commit imo.

Dudemanguy commented on video/out/vo_sdl.c:
```diff
@@ -599,7 +599,7 @@ static void wait_events(struct vo *vo, int64_t until_time_us)
         }
         case SDL_MOUSEMOTION:
-            mp_input_set_mouse_pos(vo->input_ctx, ev.motion.x, ev.motion.y);
+            mp_input_set_mouse_pos(vo->input_ctx, ev.motion.x, ev.motion.y, 0);
```
SDL should be able to handle modifiers.

occivink replied:
As far as I can see, the SDL vo does not handle keyboard modifiers at all with mouse events at the moment. 
The SDL events also do not contain any field with the modifiers held, should I query the modifier state in all these events?

Dudemanguy replied:
Does `SDL_GetModState()` not work?

Akemi commented:
> I personally don't think so. In the calc geometry case, there's value in keeping the old functions since not everything needs `vo_calc_window_geometry3`. Honestly, `vo_calc_window_geometry2` probably didn't need to be a separate function anyway and simply adding the DPI as an extra argument and passing 1 if DPI wasn't needed would have been fine but that's an old story. As for this, if the vo is not expected to handle modifiers, just passing 0 is perfectly fine imo since it's only used internally.

just that i am not the one holding anything back. i am fine with it either way. just wanted to ask since i wasn't sure if we had some kind of best practice in this case.

occivink committed:
vo_sdl: handle modifiers in mouse events

Previously there was passed as if no modifier was held

occivink committed:
command: allow passing modifiers in the `mouse` command

occivink commented:
Thank you for the review @Dudemanguy. I've just pushed the following changes:
* changed `MP_KEY_IS_MOUSE_MOVE` to account for modifiers
* Added support for modifiers in all mouse events to the SDL vo
* Added a "modifiers" argument to the `mouse` command. Since the `mp_input_get_key_from_name` function only parses entire key sequences, I had to add a new function for parsing just modifiers. I'm not too sure about that part, since the modifier parsing feels a bit redundant.

Dudemanguy commented:
You could probably modify `mp_input_get_key_from_name` to give it the ability to return only modifiers, but I would bet that would cause more problems than it's worth. The redundancy is likely safer. Overall, it looks like a pretty nice change.

avih commented:
> Motivation for this change: I have a plugin where the user can draw rectangles on the video with the mouse. I would like to make it similar to other programs (like inkscape), where CTRL can be held to preserve the aspect ratio, or SHIFT to preserve the center (or both)

Why is it not enough to register the modifier on mouse-down?

I think the commits construction could be improved: add modifiers support to SDL first in a commit, then support mouse pos modifiers in SDL together with all other VOs.

I have a concern that this will break mouse move coalescing. Currently it coalesces (drop the command) bases on whether or not the cmd is move, but the patches don't enhance it with move+specific-modifiers, so I think it could drop some move commands eventhough they're not the "same" as prior ones due to possibly different mods.

But first thing first - why is it not enough to register the mod on mouse down only?

occivink commented:
>Why is it not enough to register the modifier on mouse-down?

In inkscape, you can draw a rectangle like so:
1. select the rectangle tool
2. click somewhere on the canvas (this defines the first point) and hold the button down
3. move the cursor somewhere and release (this defines the second point)

If, between steps 2 and 3, you hold shift or ctrl it will change the way the rectangle is defined (its aspect ratio or centering). This can be done at any moment while drawing the rectangle, you don't have to have the modifier down while clicking the first time and so it's more flexible. 
I would like to achieve the same thing, and having to know which modifiers need to be pressed beforehand is not as nice.

I need to look at the mouse move coalescing, I wasn't aware of the feature.

avih commented:
> and having to know which modifiers need to be pressed beforehand is not as nice.

Sure. Personally I wouldn't consider this very specific niceness, of which you're script is likely to be the sole user of ever, enough reason to make all the changes which your patches do. But if others feel it's well worth it, so be it.

Still, my other two points hold.

occivink commented:
>of which you're script is likely to be the sole user of ever

It's not too difficult to think of other applications, for example a seekbar with chapter markers where the cursor snaps to chapter positions if (or unless) a modifier is pressed.

>all the changes which your patches do

well initially the change was mostly about one function, `mp_input_set_mouse_pos`, it just so happens that some tangentially relevant changes (which I think make sense) were requested. I don't have a use for them myself, so I could also split or drop them if you prefer.

I've reordered the commits as you requested. As for the coalescing of events, I'm afraid `mp_cmd` does not contain enough information at this point to know whether both events had the same modifiers, so I'm not sure how to approach such a change. 
But thinking some more about it, applications of `MOUSE_MOVE` seem to care not about all events, but only about the latest state (unlike a button/key press, where you care about all of them). This is probably why this coalescing exists in the first place, and why it doesn't consider x or y position when merging them. This also makes me think that it's ok to coalesce mouse events, even if they have different modifiers.

avih commented:
> I've reordered the commits as you requested.

Looks good. Thanks.

> As for the coalescing of events, I'm afraid `mp_cmd` does not contain enough information at this point

Indeed.

> But thinking some more about it, applications of `MOUSE_MOVE` seem to care not about all events, but only about the latest state

It's not that they don't care per-se, it's just that mpv "eats" consecutive move events which were not yet consumed. But with this PR, which supports mods during move - they can get lost.

occivink commented:
>But with this PR, which supports mods during move - they can get lost.

Yes I realized, but this is only the "old" mouse move event that gets lost. Similarly, if two consecutive events with different x/y coordinates are fired they might get coalesced and the earlier one be lost, even though they were not the same. It seems to me like this is the same with the modifiers.

avih commented:
> It seems to me like this is the same with the modifiers

I don't think so. Imagine that you move the mouse and then press and release  modifier - it can get lost. Might not matter in your use case, but could matter to others.

avih commented:
Also, your script could implement many "modifiers" for mouse using plain keys (which you can observe both down and up events). Might not be as nice as "normal" modifiers, but would work equally well during the move. And these will not get lost.

occivink commented:
>someone might want to press a modifier during move to "mark" positions during the move

This is not what modifiers are used for. You're describing a discrete event, which is clearly the role of a button/key press. You speak of hypothetical usecases (even though earlier there were likely not going to be any other ever...) that can break due to coalescing, but can you describe a realistic one?

Actually the current coalescing is already broken according to your logic. Imagine a script that detects mouse gestures, a circular motion could be indistinguishable from no movement if enough events are coalesced. But that likely does not happen, the logic is there not to flood the commands queue when 1+khz refresh rate mouse are used.

Of course if you just want to close this, feel free.

avih commented:
Move events are semi continuous because they represent an analogue motion (of the mouse) and indeed they can come from 1KHz mouse. So they're produced at the consumption rate by dropping prior moves which were not consomed.

In contrast, modifiers up/down ARE indeed discrete key events which accompany the move. You can't coalesce moves with different modifiers and call it reasonable IMO.

occivink committed:
input: record modifiers of mouse_move commands

To ensure that we do not coalesce mouse_move events which have
different modifiers.

occivink commented:
>You can't coalesce moves with different modifiers and call it reasonable IMO.

It seems reasonable to me as long as I haven't seen a case where it could matter. You didn't address the potential motion detection failure either. To me they are the same symptoms, and "it's a bug", just like you say.

Anyway, I've pushed a commit that adds a field to `mp_cmd` for this specific purpose. This seems like an overkill solution to a theoretical problem and I doubt that this will be accepted. If it does, I'll squash it with the other input commit.
If you have a better approach in mind, please let me know.

avih commented:
As I said earlier, I'm not in favor of adding this considering the available alternatives and value vs amount of code and potential issues, so I'm not trying to figure out which approach is best. I was merely mentioning an issue which apparently no one was aware of.

occivink commented:
Ok. I hope you realize that it's frustrating and a waste of time (for both of us) to have you request justifications, usecases and changes, only to later say that you don't care and would prefer to just reject the feature altogether.

Dudemanguy commented:
For what it's worth, the commit that handles coalescing looks totally reasonable to me, and I think it makes sense to include it in this PR.

avih commented:
I still think that touching 15 files (excluding the 1st commit of SDL fix), including all VOs, duplicating key parsing code and adding complexity and potential for issues, is not worth the additional functionality of being able to change modifiers during move instead of simply registering it on mouse down - which already works without this PR.

occivink commented:
As the name, documentation and your previous wording suggest, the "mouse-pos" property represents the "last known **position**" of the mouse, so there's no incompleteness here.
What you're proposing a separate enhancement. I suggest that you make a separate pull request if this one is accepted, this one has already increased in scope. Don't forget to propose plenty of usecases to justify this new enhancement.

avih commented:
> As the name, documentation and your previous wording suggest, the "mouse-pos" property represents the "last known **position**" of the mouse, so there's no incompleteness here.

There WAS no incompleteness, before every other place which refers to position now also refers to modifiers.

Dudemanguy commented:
I don't really understand why `mouse-pos` would hypothetically return anything other than the mouse position.

Dudemanguy commented:
Closing some very old PRs that seemed to stall out. Feel free to complain if you intend to work on it.
